// firestore-storage-core/dist/lib/storage/index_manger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryScope = exports.FieldArrayConfig = exports.IndexFieldOrder = exports.IndexManager = void 0;
class IndexManager {
    constructor() {
        this.indexes = [];
        this.fieldOverrides = [];
    }
    addIndex(collectionGroup, queryScope) {
        return new IndexBuilder(this, collectionGroup, queryScope);
    }
    addOverride(collectionGroup, fieldPath) {
        return new FieldOverrideBuilder(this, collectionGroup, fieldPath);
    }
    toObject() {
        return {
            indexes: this.indexes,
            fieldOverrides: this.fieldOverrides
        };
    }
    toJSON(space) {
        return JSON.stringify(this.toObject(), null, space);
    }
}
exports.IndexManager = IndexManager;
class IndexBuilder {
    constructor(parent, collectionGroup, queryScope) {
        this.parent = parent;
        this.entry = {
            collectionGroup,
            queryScope,
            fields: []
        };
    }
    field(fieldPath, order) {
        this.entry.fields.push({
            fieldPath: fieldPath,
            order: order || IndexFieldOrder.Asc
        });
        return this;
    }
    arrayField(fieldPath, config) {
        this.entry.fields.push({
            fieldPath: fieldPath,
            arrayConfig: config
        });
        return this;
    }
    add() {
        if (this.entry.fields.length < 2) {
            throw new Error(`Not enough fields provided to create an index for ${this.entry.collectionGroup}`);
        }
        this.parent.indexes.push(this.entry);
        return this.parent;
    }
}
class FieldOverrideBuilder {
    constructor(parent, collectionGroup, fieldPath) {
        this.parent = parent;
        this.entry = {
            collectionGroup, fieldPath,
            indexes: []
        };
    }
    order(queryScope, order) {
        this.entry.indexes.push({ queryScope, order });
        return this;
    }
    array(queryScope, arrayConfig) {
        this.entry.indexes.push({ queryScope, arrayConfig });
        return this;
    }
    add() {
        this.parent.fieldOverrides.push(this.entry);
        return this.parent;
    }
}
var IndexFieldOrder;
(function (IndexFieldOrder) {
    IndexFieldOrder["Asc"] = "ASCENDING";
    IndexFieldOrder["Desc"] = "DESCENDING";
})(IndexFieldOrder = exports.IndexFieldOrder || (exports.IndexFieldOrder = {}));
var FieldArrayConfig;
(function (FieldArrayConfig) {
    FieldArrayConfig["Contains"] = "CONTAINS";
})(FieldArrayConfig = exports.FieldArrayConfig || (exports.FieldArrayConfig = {}));
var QueryScope;
(function (QueryScope) {
    QueryScope["Collection"] = "COLLECTION";
    QueryScope["CollectionGroup"] = "COLLECTION_GROUP";
})(QueryScope = exports.QueryScope || (exports.QueryScope = {}));
