// ff-utils/dist/lib/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const lodash_isplainobject_1 = require("lodash.isplainobject");
const lodash_camelcase_1 = require("lodash.camelcase");
function camelCaseObjectKeys(object) {
    const obj = {};
    Object.keys(object).forEach((key) => {
        const ccKey = lodash_camelcase_1.default(key);
        const value = object[key];
        if (lodash_isplainobject_1.default(value)) {
            obj[ccKey] = this.camelCaseObjectKeys(value);
        }
        else {
            obj[ccKey] = value;
        }
    });
    return obj;
}
exports.camelCaseObjectKeys = camelCaseObjectKeys;
function round(num, precision) {
    precision = precision || 0;
    const p = Math.pow(10, precision);
    return Math.round(num * p) / p;
}
exports.round = round;
function random(lower, upper) {
    lower = lower || 0;
    upper = upper || 1;
    return Math.round(lower + Math.random() * (upper - lower));
}
exports.random = random;
function formatBytes(bytes, decimals = 2) {
    if (bytes === 0)
        return '0 B';
    const k = 1024;
    const dm = decimals < 0 ? 0 : decimals;
    const sizes = ['B', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));
    return parseFloat((bytes / Math.pow(k, i)).toFixed(dm)) + ' ' + sizes[i];
}
exports.formatBytes = formatBytes;
function wait(ms) {
    return new Promise((resolve) => {
        setTimeout(resolve, ms);
    });
}
exports.wait = wait;
function debounce(func, ms) {
    let timeout;
    let lastCall;
    let lastArgs;
    return function (...a) {
        const context = this;
        lastArgs = arguments;
        const later = function () {
            timeout = null;
            lastCall = Date.now();
            func.apply(context, lastArgs);
        };
        clearTimeout(timeout);
        if (lastCall && lastCall + ms > Date.now()) {
            timeout = setTimeout(later, ms);
        }
        else {
            later();
        }
    };
}
exports.debounce = debounce;
function enumToArray(Enum) {
    let keys = Object.keys(Enum);
    const hasNumbers = keys.some(k => !isNaN(Number(k)));
    if (hasNumbers) {
        keys = keys.filter(v => isNaN(Number(v)));
    }
    return keys.map(key => (Enum[key]));
}
exports.enumToArray = enumToArray;
