// firestore-storage/dist/lib/storage/accumulator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StorageEventAccumulator = void 0;
const firestore_storage_1 = require("./firestore_storage");
class StorageEventAccumulator {
    constructor(emitter) {
        this.emitter = emitter;
        this.events = {};
        this.trackerFunc = this.track.bind(this);
        emitter.on(firestore_storage_1.StorageEventType.Read, this.trackerFunc);
        emitter.on(firestore_storage_1.StorageEventType.Write, this.trackerFunc);
        emitter.on(firestore_storage_1.StorageEventType.Delete, this.trackerFunc);
    }
    track(event) {
        this.events[event.type] = this.events[event.type] || {};
        this.events[event.type][event.collection] = this.events[event.type][event.collection] || 0;
        this.events[event.type][event.collection] += event.count;
    }
    off() {
        this.emitter.off(firestore_storage_1.StorageEventType.Read, this.trackerFunc);
        this.emitter.off(firestore_storage_1.StorageEventType.Write, this.trackerFunc);
        this.emitter.off(firestore_storage_1.StorageEventType.Delete, this.trackerFunc);
    }
    reset() {
        this.events = {};
    }
    get() {
        return this.events;
    }
}
exports.StorageEventAccumulator = StorageEventAccumulator;
